src/test/java/BaseTest.java

```diff
@@ -1,11 +1,109 @@
+import org.openqa.selenium.firefox.FirefoxDriver;
+import org.openqa.selenium.interactions.Actions;
+import org.openqa.selenium.remote.DesiredCapabilities;
+import org.openqa.selenium.remote.RemoteWebDriver;
+import org.openqa.selenium.support.ui.FluentWait;
+import org.openqa.selenium.support.ui.Wait;
+import org.openqa.selenium.support.ui.WebDriverWait;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
 import org.testng.annotations.BeforeSuite;
+import org.testng.annotations.Parameters;
+
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.time.Duration;
 
 public class BaseTest {
+    public static WebDriver driver = null;
+    //public ChromeOptions options = new ChromeOptions();
+    public WebDriverWait wait;
+    public Wait<WebDriver>fluentWait;
+    public Actions actions;
+
+    //public String url = "https://qa.koel.app/";
 
     @BeforeSuite
     static void setupClass() {
-        WebDriverManager.chromedriver().setup();
+        //WebDriverManager.chromedriver().setup();
+        WebDriverManager.firefoxdriver().setup();
+    }
+
+    @BeforeMethod
+    @Parameters({"baseURL"})
+    public void launchBrowser (String baseURL) throws MalformedURLException{
+        //pre-condition
+
+        driver=pickBrowser(System.getProperty("browser"));
+        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(10));
+        driver.manage().window().maximize();
+        wait = new WebDriverWait(driver,Duration.ofSeconds(10));
+        fluentWait = new FluentWait<>(driver)
+                .withTimeout(Duration.ofSeconds(30));
+        actions = new Actions(driver);
+        driver.get(baseURL);
+
+    }
+    public static WebDriver pickBrowser(String browser) throws MalformedURLException {
+        DesiredCapabilities caps = new DesiredCapabilities();
+        String gridURL = "http://192.168.1.203:4444";
+        switch (browser){
+            case "firefox":
+                WebDriverManager.firefoxdriver().setup();
+                return driver = new FirefoxDriver();
+            case "MicrosoftEdge":
+                WebDriverManager.edgedriver().setup();
+                EdgeOptions edgeOptions = new EdgeOptions();
+                edgeOptions.addArguments("--remote-allow-origins=*");
+                return driver = new EdgeDriver(edgeOptions);
+
+            //Grid Cases
+            case "grid-edge":
+                caps.setCapability("browserName","MicrosoftEdge");
+                return driver = new RemoteWebDriver(URI.create(gridURL).toURL(),caps);
+
+            case "grid-firefox":
+                caps.setCapability("browserName","Firefox");
+                return driver = new RemoteWebDriver(URI.create(gridURL).toURL(),caps);
+
+            case "grid-chrome":
+                caps.setCapability("browserName","Chrome");
+                return driver = new RemoteWebDriver(URI.create(gridURL).toURL(),caps);
+
+
+
+
+
```

Comment:
**Please delete the empty line!** An Empty Line is a "Code Smell"
The short answer is, _They’re more an annoyance than a problem.They won’t actually break anything—which is why your code continues to run—but they make it awkward to read and change your code, especially if you’re dealing with multiple line breaks per line of real code. They also add just a tiny bit of size to your files, which is nice to reduce._